File: backend/app/models/dongpa.py
"""Pydantic models for 동파법 (Dongpa) backtest API.

QQQ 주봉 기반 안전/공세 모드 전환으로 SOXL을 매매하는 전략.
"""

from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.models.validators import PositiveCapitalValidator


class DongpaModeRule(BaseModel):
    """모드 전환 규칙 (QQQ 주봉 기반)."""

    indicator: str  # 'qqq_ma_cross', 'qqq_trend_deviation', 'qqq_weekly_roc', 'qqq_ma_slope'
    operator: Literal["gt", "gte", "lt", "lte", "eq", "neq"]
    value: float | str

    model_config = {"populate_by_name": True}


class DongpaMode(BaseModel):
    """안전/공세 모드 정의."""

    mode_id: str = Field(alias="modeId")  # 'safe', 'aggressive'
    name: str  # '안전', '공세'
    strategy_id: str = Field(alias="strategyId")
    parameters: dict[str, float | int | str | bool]

    model_config = {"populate_by_name": True}


class DongpaTrigger(BaseModel):
    """모드 전환 트리거."""

    id: str
    name: str
    conditions: list[DongpaModeRule]
    logic: Literal["and", "or"] = "and"
    target_mode_id: str = Field(alias="targetModeId")
    priority: int = 0

    model_config = {"populate_by_name": True}


class DongpaConfig(BaseModel):
    """동파법 설정."""

    modes: list[DongpaMode]
    triggers: list[DongpaTrigger]
    default_mode_id: str = Field(default="safe", alias="defaultModeId")
    qqq_ma_period: int = Field(default=10, ge=2, le=52, alias="qqqMaPeriod")  # QQQ 주봉 MA 기간
    position_handling: Literal["carry", "liquidate"] = Field(
        default="carry", alias="positionHandling"
    )

    model_config = {"populate_by_name": True}


class DongpaModeEvent(BaseModel):
    """모드 전환 이벤트."""

    time: int
    day_index: int = Field(alias="dayIndex")
    from_mode_id: str = Field(alias="fromModeId")
    to_mode_id: str = Field(alias="toModeId")
    trigger_id: str = Field(alias="triggerId")
    trigger_name: str = Field(alias="triggerName")
    qqq_indicators: dict[str, float | str | None] = Field(alias="qqqIndicators")
    portfolio_value: float = Field(alias="portfolioValue")
    open_tiers: int = Field(alias="openTiers")

    model_config = {"populate_by_name": True}


class DongpaSegmentResult(BaseModel):
    """모드 구간별 성과."""

    mode_id: str = Field(alias="modeId")
    mode_name: str = Field(alias="modeName")
    strategy_id: str = Field(alias="strategyId")
    start_time: int = Field(alias="startTime")
    end_time: int = Field(alias="endTime")
    start_value: float = Field(alias="startValue")
    end_value: float = Field(alias="endValue")
    segment_return: float = Field(alias="segmentReturn")
    segment_mdd: float = Field(alias="segmentMdd")
    trade_count: int = Field(alias="tradeCount")
    candle_count: int = Field(alias="candleCount")

    model_config = {"populate_by_name": True}


class DongpaBacktestRequest(BaseModel, PositiveCapitalValidator):
    """동파법 백테스트 요청."""

    ticker_id: str = Field(default="SOXL", alias="tickerId")
    start_date: str | None = Field(default=None, alias="startDate")
    end_date: str | None = Field(default=None, alias="endDate")
    initial_capital: float = Field(alias="initialCapital")
    apply_fee: bool = Field(default=True, alias="applyFee")
    market: str = Field(default="us")
    dongpa_config: DongpaConfig = Field(alias="dongpaConfig")

    model_config = {"populate_by_name": True}


class DongpaBacktestResponse(BaseModel):
    """동파법 백테스트 응답."""

    trades: list  # Trade objects
    equity: list  # EquityPoint objects
    cash: list  # EquityPoint objects
    metrics: dict  # PerformanceMetrics
    price_data: list = Field(default_factory=list, alias="priceData")
    execution_time: float = Field(alias="executionTime")
    mode_events: list[DongpaModeEvent] = Field(alias="modeEvents")
    segments: list[DongpaSegmentResult]
    active_mode_timeline: list[dict] = Field(alias="activeModeTimeline")
    qqq_weekly_data: list[dict] = Field(default_factory=list, alias="qqqWeeklyData")

    model_config = {"populate_by_name": True}


# --- 동파법 옵티마이저 모델 ---


class DongpaConditionRange(BaseModel):
    """조건별 값 범위 (옵티마이저용)."""

    trigger_index: int = Field(alias="triggerIndex")
    condition_index: int = Field(alias="conditionIndex")
    indicator: str
    operator: str
    min_value: float | None = Field(default=None, alias="minValue")
    max_value: float | None = Field(default=None, alias="maxValue")
    step: float | None = None

    model_config = {"populate_by_name": True}


class DongpaParameterRange(BaseModel):
    """단일 파라미터 범위."""

    min: float
    max: float
    step: float

    model_config = {"populate_by_name": True}


class DongpaModeParameterRange(BaseModel):
    """모드별 전략 파라미터 범위."""

    mode_id: str = Field(alias="modeId")
    drop_percent: DongpaParameterRange | None = Field(default=None, alias="dropPercent")
    target_profit: DongpaParameterRange | None = Field(default=None, alias="targetProfit")
    stop_loss_days: DongpaParameterRange | None = Field(default=None, alias="stopLossDays")

    model_config = {"populate_by_name": True}


class DongpaOptimizerRequest(BaseModel, PositiveCapitalValidator):
    """동파법 옵티마이저 요청."""

    ticker_id: str = Field(default="SOXL", alias="tickerId")
    start_date: str | None = Field(default=None, alias="startDate")
    end_date: str | None = Field(default=None, alias="endDate")
    initial_capital: float = Field(alias="initialCapital")
    apply_fee: bool = Field(default=True, alias="applyFee")
    market: str = Field(default="us")
    dongpa_config: DongpaConfig = Field(alias="dongpaConfig")
    condition_ranges: list[DongpaConditionRange] = Field(alias="conditionRanges")
    top_n: int = Field(default=50, alias="topN")
    algorithm: Literal["grid_search", "monte_carlo"] = Field(default="grid_search")
    mode_parameter_ranges: list[DongpaModeParameterRange] | None = Field(
        default=None, alias="modeParameterRanges"
    )

    model_config = {"populate_by_name": True}


class DongpaOptimizerResultItem(BaseModel):
    """동파법 옵티마이저 결과 항목."""

    rank: int
    condition_values: list[dict] = Field(alias="conditionValues")
    total_return: float = Field(alias="totalReturn")
    cagr: float
    mdd: float
    sharpe_ratio: float = Field(alias="sharpeRatio")
    win_rate: float = Field(alias="winRate")
    total_trades: int = Field(alias="totalTrades")
    switch_count: int = Field(alias="switchCount")
    composite_score: float = Field(alias="compositeScore")

    model_config = {"populate_by_name": True}

File: frontend/types/dongpa.ts
import { OHLCV, PerformanceMetrics, Trade } from './backtest';

// QQQ 주봉 기반 모드 전환 규칙
export interface DongpaModeRule {
  indicator: string; // 'qqq_ma_cross', 'qqq_trend_deviation', 'qqq_weekly_roc', 'qqq_ma_slope'
  operator: 'eq' | 'neq' | 'gt' | 'gte' | 'lt' | 'lte';
  value: number | string;
}

// 안전/공세 모드 정의
export interface DongpaMode {
  modeId: string; // 'safe', 'aggressive'
  name: string;
  strategyId: string;
  parameters: Record<string, number | string | boolean>;
}

// 모드 전환 트리거
export interface DongpaTrigger {
  id: string;
  name: string;
  conditions: DongpaModeRule[];
  logic: 'and' | 'or';
  targetModeId: string;
  priority: number;
}

// 동파법 설정
export interface DongpaConfig {
  modes: DongpaMode[];
  triggers: DongpaTrigger[];
  defaultModeId: string;
  qqqMaPeriod: number;
  positionHandling: 'carry' | 'liquidate';
}

// 모드 전환 이벤트
export interface DongpaModeEvent {
  time: number;
  dayIndex: number;
  fromModeId: string;
  toModeId: string;
  triggerId: string;
  triggerName: string;
  qqqIndicators: Record<string, number | string | null>;
  portfolioValue: number;
  openTiers: number;
}

// 모드 구간별 성과
export interface DongpaSegmentResult {
  modeId: string;
  modeName: string;
  strategyId: string;
  startTime: number;
  endTime: number;
  startValue: number;
  endValue: number;
  segmentReturn: number;
  segmentMdd: number;
  tradeCount: number;
  candleCount: number;
}

// QQQ 주봉 차트 데이터
export interface QQQWeeklyPoint {
  date: string;
  close: number;
  ma: number | null;
  maCross: string;
  trendDeviation: number;
  weeklyRoc: number;
  maSlope: number;
}

// 백테스트 요청/응답
export interface DongpaBacktestRequest {
  tickerId: string;
  startDate?: string;
  endDate?: string;
  initialCapital: number;
  applyFee?: boolean;
  market?: string;
  dongpaConfig: DongpaConfig;
}

export interface DongpaBacktestResponse {
  trades: Trade[];
  equity: { time: number; value: number }[];
  cash: { time: number; value: number }[];
  metrics: PerformanceMetrics;
  priceData: OHLCV[];
  executionTime: number;
  modeEvents: DongpaModeEvent[];
  segments: DongpaSegmentResult[];
  activeModeTimeline: { time: number; modeId: string }[];
  qqqWeeklyData: QQQWeeklyPoint[];
}

// 지표 정보
export interface DongpaIndicatorInfo {
  id: string;
  label: string;
  type: 'string' | 'number';
  operators: string[];
  values?: string[];
  unit?: string;
}

// 모드 색상
export const MODE_COLORS = {
  safe: {
    bg: 'bg-blue-50 dark:bg-blue-900/20',
    border: 'border-blue-200 dark:border-blue-800',
    text: 'text-blue-700 dark:text-blue-300',
    badge: 'bg-blue-100 dark:bg-blue-900/40 text-blue-800 dark:text-blue-200',
  },
  aggressive: {
    bg: 'bg-red-50 dark:bg-red-900/20',
    border: 'border-red-200 dark:border-red-800',
    text: 'text-red-700 dark:text-red-300',
    badge: 'bg-red-100 dark:bg-red-900/40 text-red-800 dark:text-red-200',
  },
} as const;

export type ModeId = keyof typeof MODE_COLORS;

// 전략 옵션 (스위칭과 공유)
export const STRATEGY_OPTIONS = [
  { id: 'ddeolsapro1', name: '떨사프로1', ratios: '5/10/15/20/25/25%' },
  { id: 'ddeolsapro2', name: '떨사프로2', ratios: '10/15/20/25/20/10%' },
  { id: 'ddeolsapro_custom', name: '떨사 커스텀', ratios: '사용자 정의' },
] as const;

export const STRATEGY_DEFAULTS: Record<string, Record<string, number | string | boolean>> = {
  ddeolsapro1: { dropPercent: 0.01, targetProfit: 0.01, stopLossDays: 10, stopLossDayBuy: 'allow' },
  ddeolsapro2: { dropPercent: 0.01, targetProfit: 1.5, stopLossDays: 10, stopLossDayBuy: 'allow' },
  ddeolsapro_custom: { dropPercent: 0.01, targetProfit: 0.01, stopLossDays: 10, stopLossDayBuy: 'allow', divisions: 6 },
};

// 연산자 라벨
export const OPERATOR_LABELS: Record<string, string> = {
  eq: '==',
  neq: '!=',
  gt: '>',
  gte: '>=',
  lt: '<',
  lte: '<=',
};

// ===== 옵티마이저 =====

export interface DongpaConditionRange {
  triggerIndex: number;
  conditionIndex: number;
  indicator: string;
  operator: string;
  minValue?: number;
  maxValue?: number;
  step?: number;
}

export interface DongpaOptimizerRequest {
  tickerId: string;
  startDate?: string;
  endDate?: string;
  initialCapital: number;
  applyFee?: boolean;
  market?: string;
  dongpaConfig: DongpaConfig;
  conditionRanges: DongpaConditionRange[];
  topN?: number;
  algorithm?: 'grid_search' | 'monte_carlo';
  modeParameterRanges?: DongpaModeParameterRange[];
}

export interface DongpaModeParameterRange {
  modeId: string;
  dropPercent?: { min: number; max: number; step: number };
  targetProfit?: { min: number; max: number; step: number };
  stopLossDays?: { min: number; max: number; step: number };
}

export interface DongpaOptimizerResultItem {
  rank: number;
  conditionValues: { triggerIndex: number; conditionIndex: number; value: number | string }[];
  totalReturn: number;
  cagr: number;
  mdd: number;
  sharpeRatio: number;
  winRate: number;
  totalTrades: number;
  switchCount: number;
  compositeScore: number;
}

export interface DongpaOptimizerCallbacks {
  onStart?: (start: { jobId: string }) => void;
  onProgress?: (progress: { completed: number; total: number; percent: number }) => void;
  onResults?: (results: DongpaOptimizerResultItem[]) => void;
  onDone?: (done: { totalTime: number; resultsCount: number }) => void;
  onCancelled?: (cancelled: { jobId: string; completed: number; total: number }) => void;
  onError?: (error: Error) => void;
}
